staging/src/k8s.io/apiserver/pkg/endpoints/installer.go

```diff
@@ -604,6 +604,26 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 		if err != nil {
 			return nil, nil, fmt.Errorf("failed to create field manager: %v", err)
 		}
+
+		if subresource == "scale" {
+			parentResourceGVK, err := GetResourceKind(a.group.GroupVersion, a.group.Storage[resource], a.group.Typer)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			reqScope.ParentFieldManager, err = fieldmanager.NewDefaultFieldManager(
```

Comment:
I typically try to avoid having variables that only make sense in a specific context and not others. In that case, the ParentFieldManager member is always in scopes but is only used for specific sub-resource. In that case (scale sub-resource), do we actually use both fieldmanagers? Is it possible that the fieldmanager above is misconfigured?

Reply:
I haven't looked extensively, but it's possible that we actually run the FieldManager on the scale object itself before :-(.

Reply:
It should be quite safe, `options` is created [here](https://github.com/kubernetes/kubernetes/pull/95921/files#diff-c4afbb7d6a8bdb1d6475f18abc92caa6c413f879aeeb189187c569f3f6c146f4R584). Do you prefer me to add extra checks?

Reply:
Yes, if I am not mistaken, it's used on the scale object itself too (e.g. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L553)

Reply:
Yeah, for nothing, but yes :D

Reply:
Yeah, that's a good point, thanks for mentioning that.

@nodo, can we start by adding a test that reproduces that problem. That is:
- create a deployment,
- Run "Apply" with a "Scale" object that has a different number for replicas, without "force", and make sure we have a conflict.

My understanding is that 2. is mostly what we already do, we need to implement 1. to fix the test described above?

Reply:
Right, I have added the test here: https://github.com/kubernetes/kubernetes/pull/95921/commits/2c1408803f8d33cfbf2d331ac594380e788f6795
and it works as expected.

If I understand correctly. The field manager of the main resource is invoked also when running "apply on "scale" so it triggers a conflict. Am I missing something though?

Reply:
Ah! There's something tricky happening here. To verify that, I think you could write a test where apply will actually ALWAYS conflict on scale. When you apply an object that doesn't have a managedFields set (and that's the case of EVERY scale objects here), we create one that we call "before-first-apply" with existing fields. Here, you're getting a conflict with that manager, not the actual manager that changed the value. Every time you apply, we'll re-create this "before-first-apply" manager (even though you're supposed to own the field), and you'll get a conflict.

I don't know yet how to solve that.

Reply:
Thanks a lot for spotting, that's quite a tricky edge case :)

I think a possible solution would be to not run `Apply` on scale objects. E.g. before [this line](https://github.com/nodo/kubernetes/blob/ssa-scale-ownership-deployment/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L441)

Add something like this:

```go
if scale, ok := patchObject.(*autoscaling.Scale); ok {
  return scale, nil
}
```
Applying the field manager here doesn't have any effect anyway.

I tried to implement it but I got stuck because I couldn't find a way to convert an `*unstructured.Unstructured` to `*autoscaling.Scale` and I wanted to check with you first.

What do you think about this solution? If it's a good way forward, how to do the conversion?

Reply:
I have done some more tests and I was trying to convert the `Unstructured` object using the following code:

```go
scaleGVK := schema.GroupVersionKind{Group: "autoscaling", Version: "v1", Kind: "Scale"}
if p.kind == scaleGVK {
        var scale autoscaling.Scale
        err := runtime.DefaultUnstructuredConverter.FromUnstructured(patchObj.UnstructuredContent(), &scale)
        return &scale, err
}
```

This doesn't work for a reason I don't fully understand. The resulting `scale` object doesn't have any metadata, and I am getting this error:
```
error the name of the object (deployment based on URL) was undeterminable: name must be provided
```

Digging a bit deeper, I have noticed that [the Scale object](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/autoscaling/types.go#L28-L41) does not specify any JSON annotations, why is that? Other types such as [Deployment](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/apps/v1/types.go#L254-L267) have them. Could this be a problem?

Reply:
> Digging a bit deeper, I have noticed that the Scale object does not specify any JSON annotations, do you know the reason for that? Other types such as Deployment have them. Could this be a problem?

You're looking at the "internal type" rather than the "external type". Internal types are not converted to json and hence don't have the tags. You should be looking at the external type (note the difference in the URL, it's probably too subtle ...): https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/autoscaling/v1/types.go#L114-L127